# Στοχαστικές Διεργασίες και Βελτιστοποίηση στη Μηχανική Μάθηση

## 6ο Εργαστήριο - *Restricted Boltzmann Machine και Deep Belief Network*

- Ονομ/νυμο: Χρήστος Νίκου
- AM: 03400146
- Ιδιότητα: Μεταπτυχιακός φοιτητής Επιστήμης Δεδομένων και Μηχανικής Μάθησης (ΕΔΕΜΜ)
- Ηλεκτρονική Διεύθυνση: christosnikou@mail.ntua.gr / chrisnick92@gmail.com

<h1><b>Restricted Boltzmann Machine και Deep Belief Network</b></h1>
<p align="justify">Στην συγκεκριμένη άσκηση θα μελετήσετε τον τρόπο λειτουργίας μιας <i>RBM (<a href="https://en.wikipedia.org/wiki/Restricted_Boltzmann_machine">Restricted Boltzmann Machine</a>)</i> καθώς και των <i>DBN (<a href="https://en.wikipedia.org/wiki/Deep_belief_network">Deep Belief Network</a>)</i>, χρησιμοποιώντας το έτοιμο πρόγραμμα που σας δίνεται.Το συγκεκριμένο πρόγραμμα αξιοποιεί το <a href="https://en.wikipedia.org/wiki/MNIST_database">dataset του <i>MNIST</i></a>, όπου είναι μια μεγάλη βάση δεδομένων με χειρόγραφα ψηφία που χρησιμοποιείται συνήθως για την εκπαίδευση διαφόρων συστημάτων επεξεργασίας εικόνας. Για την άσκηση, θα πρέπει να χρησιμοποιήσετε το αρχείο <i>mnist_original.mat</i>, το οποίο είναι διαθέσιμο από <a href="https://www.kaggle.com/datasets/avnishnish/mnist-original?resource=download">εδώ</a>.</p>
<p align="justify">Μία αρκετά σημαντική εφαρμογή της <i>RBM</i> είναι η εξαγωγή χαρακτηριστικών (feature representation) από ένα dataset με σκοπό την αναπαράσταση της εισόδου (ορατοί νευρώνες) με ένα διάνυσμα μικρότερης διάστασης (κρυφοί νευρώνες). Στη συγκεκριμένη άσκηση θα συγκρίνετε την ακρίβεια ενός ταξινομητή ψηφίων με τη χρήση του αλγορίθμου <i>Logistic Regression</i>, όταν εκείνος δέχεται ως είσοδο το dataset (i) χωρίς να έχει υποστεί επεξεργασία από το <i>RBM</i>, (ii) αφου υποστεί επεξεργασία από το <i>RBM</i>, (iii) με τη χρήση <i>DBN</i>, δηλαδή δύο stacked <i>RBM</i>.</p>
<p align="justify"> Με βάση τον κώδικα που σας έχει δοθεί, καλείστε να απαντήσετε στα παρακάτω ερωτήματα:</p>
<ul>
<li>Να περιγράψετε σύντομα τον τρόπο λειτουργίας μιας <i>RBM</i>. Τι διαφορές έχει σε σχέση με μία <i> Μηχανή Boltzmann</i>;</li>
<li>Ποια είναι η λογική των <i>DBN</i> και σε τι προβλήματα τα αξιοποιούμε;</li>
<li>Να αναφέρετε τις βασικότερες εφαρμογές των <i>RBM</i> και <i>DBN</i>.</li>
<li>Εκτός από <i>RBM</i>, τι άλλα μοντέλα μπορούν να χρησιμοποιηθούν για να δημιουργήσουν <i>DBN</i>.</li>
<li>Συγκρίνετε τα αποτελέσματα της ταξινόμησης με τον αλγόριθμo <i>Logistic Regression</i> χωρίς τη χρήση <i>RBM</i> σε σχέση με τα αποτελέσματα της ταξινόμησης που έχει χρησιμοποιηθεί η <i>RBM</i> καθώς και με αυτή όπου χρησιμοποιούνται <i>RBM</i> και <i>DBN</i> για την εξαγωγή των χαρακτηριστικών. Τι παρατηρείτε ως προς την ακρίβεια των αποτελεσμάτων;</li>
</ul>

><b>1.</b> *Να περιγράψετε σύντομα τον τρόπο λειτουργίας μιας RBM. Τι διαφορές έχει σε σχέση με μία Μηχανή Boltzmann;*


Ένα <b>Restricted Boltzmann Machine (RBM)</b> είναι ένα generative στοχαστικό νευρωνικό δίκτυο το οποίο μαθαίνει την κατανομή των δεδομένων που του δίνονται. Μερικές απ'τις εφαρμογές των *RBM* είναι οι εξής:

- Μείωση διαστατικότητας (*dimensionality reduction*)
- Εκμάθηση χαρακτηριστικών (*feature learning*)
- Ταξινόμηση (*classification*) και
- Παλινδρόμηση (*regression*)

Η διαφορά σε σχέση με μια *Μηχανή Boltzmann* είναι ότι δεν υπάρχουν συνδέσεις μεταξύ των δύο «ομάδων» νευρώνων. Δηλαδή, δεν υπάρχουν συνδέσεις μεταξύ των ορατών (*visible*) νευρώνων όπως δεν υπάρχουν συνδέσεις και μεταξύ δύο νευρώνων που ανήκουν στους κρυφούς νευρώνες (*hidden*). Οι συνδέσεις που υπάρχουν είναι μόνο μεταξύ νευρώνων που ανήκουν σε διαφορετικές κατηγορίες. Η διαφορά αυτή των *RBM* σε σχέση με τις *Μηχανές Boltzmann* επιτρέπει στα *RBM* να αξιοποιούν μεθόδους κατάβασης κλίσεως (*gradient based methods*).

Πιο συγκεκριμένα, η εκπαίδευση μιας *RBM* αποτελείται από δύα στάδια. Το *forward pass* από το *layer* των φανερών νευρώνων προς το *layer* των κρυφών νευρώνων και το *backward pass* ακολουθώντας την αντίστροφη πορεία. Το γεγονός ότι οι *RBM* ανήκουν στην κατηγορία των μη επιβλεπόμενων μοντέλων σημαίνει δεν υπάρχει κάποια τιμή αναφοράς για να υπολογιστεί κάποιο σφάλμα μεταξύ της εξόδου του μοντέλου και της τιμής, αλλά αντίθετα, η *RBM* προσπαθεί προσεγγίσει την κατανομή των δεδομένων εισόδου $p(x)$ μέσω μιας ανακατασκευασμένης κατανομής $q(x)$. Η προσέγγιση αυτή γίνεται ελαχιστοποιώντας την απόκλιση *Kullback - Liebler* $D_{\text{KL}}(P\, ||\, Q)$ της κατανομής $q$ από την $p$ η οποία δίνεται μέσω της ποσότητας

$$D_{\text{KL}}(P\, ||\, Q) = \int p(x)\log\biggl(\frac{p(x)}{q(x)}\biggr)\,dx.$$

Ειδικότερα, η διαδικασία προσέγγισης της $p(x)$ ή αλλιώς $p(\mathbf{v})$, με $\mathbf{v}$ τα διανύσματα εισόδου, απαιτεί αρχικά των υπολογισμό των διανυσμάτων $\mathbf{h}$ στο hidden layer:

$$
\mathbf{h}^{(1)}=S\left(\mathbf{v}^{(0) T} W+\mathbf{a}\right)
$$

όπου $\mathbf{v}^{(0)}$ είναι το διάνυσμα εισόδου, $W$ τα βάρη, $\mathbf{a}$ τα biases του hidden layer και S η συνάρτηση softmax. Από τον συγκεκριμένο υπολογισμό προσεγγίζεται η κατανομή πιθανότητας:

$$
p\left(\mathbf{h}^{(1)} \mid \mathbf{v}^{(0)} ; W\right)
$$

Στη συνέχεια, υλοποιείται η φάση της ανακατασκευής του αρχικού διανύσματος:

$$
\mathbf{v}^{(1)}=S\left(\mathbf{h}^{(1) T} W+\mathbf{b}\right)
$$

όπου $\mathbf{b}$ τα biases του visible layer. Με αυτόν τον τρόπο προσεγγίζεται η κατανομή πιθανότητας:

$$
p\left(\mathbf{v}^{(1)} \mid \mathbf{h}^{(1)} ; W\right).
$$

Τώρα, για την προσέγγιση της $p(v)$ υπολογίζεται η «ενέργεια» $E(\mathbf{v},\mathbf{h})$ μεταξύ του ορατού layer και του κρυφού layer η οποία δίνεται μέσω της σχέσης

$$
E(\mathbf{v}, \mathbf{h})=-\sum_{i \in v i s i b l e} a_{i} v_{i}-\sum_{j \in \text { hidden }} b_{j} h_{j}-\sum_{i, j} v_{i} h_{j} w_{i j}
$$

όπου $v_{i}$ και $h_{j}$ είναι οι δυαδικές καταστάσεις του διανύσματος εισόδου από τον κόμβο i και του διανύσματος του hidden layer από τον κόμβο j αντίστοιχα, $a_{i}$ και $b_{j}$ είναι τα αντίστοιχα biases και $w_{ij}$ το βάρος μεταξύ τους. Τώρα, η προσέγγιση της $p(\mathbf{v})$ γίνεται μέσω της

$$
p(\mathbf{v})=\frac{\sum_{\mathbf{h}} e^{-E(\mathbf{v}, \mathbf{h})}}{\sum_{\mathbf{v}, \mathbf{h}} e^{-E(\mathbf{v}, \mathbf{h})}}.
$$



><b>2.</b> *Ποια είναι η λογική των DBN και σε τι προβλήματα τα αξιοποιούμε;*

Τα Deep Belief Networks *(DBNs)* είναι generative μοντέλα που αποτελούνται από συνδυασμό πολλαπλών Restricted Boltzmann Machines *(RBMs)*, όπου για κάθε *RBM* layer που χρησιμοποιείται, το κρυφό layer του προηγούμενου *RBM* είναι το φανερό layer του επόμενου. Μετά το τελευταίο layer μπορεί να τοποθετηθεί ένα layer το οποίο να αφορά κάποια συνάρτηση ταξινόμησης, όπως π.χ. *logistic* ή *softmax*. Έτσι, τα *DBN* πέρα από προσέγγιση κατανομών των δεδομένων μπορούν να αξιοποιηθούν και για λόγους ταξινομήσης.

><b>3 + 4.</b> *Να αναφέρετε τις βασικότερες εφαρμογές των RBM και DBN. Εκτός από RBM, τι άλλα μοντέλα μπορούν να χρησιμοποιηθούν για να δημιουργήσουν DBN.*

Μια από τις πιο βασικές εφαρμογές των *RBM* είναι η εκμάθηση της κατανομής των δεδομένων εισόδου όπως περιγράφηκε προηγουμένως. Η λογική των *RBM* βρίσκει εφαρμογές στα συστήματα που δημιουργούν προτάσεις για τους χρήστες διάφορων εφαρμογών (*recommendation systems*). Άλλες εφαρμογές έχουν να κάνουν με την εξαγωγή χρήσιμων χαρακτηριστικών και τη μείωση διαστατικότητας. Αντίστοιχα, τα *DBN* μπορούν να χρησιμοποιηθούν σε προβλήματα ταξινόμησης όπως και την αναγνώριση αντικειμένων (*object detection*). 

Άλλα μοντέλα που μπορούν να χρησιμοποιηθούν εκτός από τα *RBM* για να δημιουργήσουν *DBN* είναι οι *Autoencoders*.

><b>5.</b> *Συγκρίνετε τα αποτελέσματα της ταξινόμησης με τον αλγόριθμo Logistic Regression χωρίς τη χρήση RBM σε σχέση με τα αποτελέσματα της ταξινόμησης που έχει χρησιμοποιηθεί η RBM καθώς και με αυτή όπου χρησιμοποιούνται RBM και DBN για την εξαγωγή των χαρακτηριστικών. Τι παρατηρείτε ως προς την ακρίβεια των αποτελεσμάτων;*

In [1]:
#!/usr/bin/env python
# source: https://devdreamz.com/question/905929-stacking-rbms-to-create-deep-belief-network-in-sklearn

import numpy as np
import matplotlib.pyplot as plt

from scipy.io import loadmat
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import BernoulliRBM
from sklearn.base import clone
from sklearn.pipeline import Pipeline
from sklearn.metrics import classification_report

def norm(arr):
    arr = arr.astype(float)
    arr -= arr.min()
    arr /= arr.max()
    return arr

if __name__ == '__main__':

    # load MNIST data set
    mnist = loadmat("mnist-original.mat")
    X, Y = mnist["data"].T, mnist["label"][0]

    # normalize inputs to 0-1 range
    X = norm(X)

    # split into train, validation, and test data sets
    X_train, X_test, Y_train, Y_test = train_test_split(X,       Y,       test_size=10000, random_state=0)
    X_train, X_val,  Y_train, Y_val  = train_test_split(X_train, Y_train, test_size=10000, random_state=0)

    # --------------------------------------------------------------------------------
    # set hyperparameters

    learning_rate = 0.02 
    total_units   =  800 
    total_epochs  =  10
    batch_size    =  128

    C = 100. # optimum for benchmark model according to sklearn docs: https://scikit-learn.org/stable/auto_examples/neural_networks/plot_rbm_logistic_classification.html#sphx-glr-auto-examples-neural-networks-plot-rbm-logistic-classification-py)

    # --------------------------------------------------------------------------------
    # construct models

    # RBM
    rbm = BernoulliRBM(n_components=total_units, learning_rate=learning_rate, batch_size=batch_size, n_iter=total_epochs, verbose=1)

    # "output layer"
    logistic = LogisticRegression(C=C, solver='lbfgs', multi_class='multinomial', max_iter=200, verbose=1)

    models = []
    models.append(Pipeline(steps=[('logistic', clone(logistic))]))                                              # base model / benchmark
    models.append(Pipeline(steps=[('rbm1', clone(rbm)), ('logistic', clone(logistic))]))                        # single RBM
    models.append(Pipeline(steps=[('rbm1', clone(rbm)), ('rbm2', clone(rbm)), ('logistic', clone(logistic))]))  # RBM stack / DBN

    # --------------------------------------------------------------------------------
    # train and evaluate models

    for model in models:
        # train
        model.fit(X_train, Y_train)

        # evaluate using validation set
        print("Model performance:\n%s\n" % (
            classification_report(Y_val, model.predict(X_val))))

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
C:\Users\dem0nakos\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   14.8s finished


Model performance:
              precision    recall  f1-score   support

         0.0       0.95      0.96      0.96       995
         1.0       0.96      0.98      0.97      1121
         2.0       0.90      0.90      0.90      1015
         3.0       0.90      0.88      0.89      1033
         4.0       0.93      0.92      0.92       976
         5.0       0.90      0.88      0.89       884
         6.0       0.94      0.94      0.94       999
         7.0       0.92      0.93      0.92      1034
         8.0       0.88      0.86      0.87       923
         9.0       0.89      0.90      0.89      1020

    accuracy                           0.92     10000
   macro avg       0.92      0.91      0.92     10000
weighted avg       0.92      0.92      0.92     10000


[BernoulliRBM] Iteration 1, pseudo-likelihood = -140.98, time = 10.06s
[BernoulliRBM] Iteration 2, pseudo-likelihood = -115.93, time = 13.97s
[BernoulliRBM] Iteration 3, pseudo-likelihood = -108.03, time = 14.11s
[Bernoul

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
C:\Users\dem0nakos\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   16.7s finished


Model performance:
              precision    recall  f1-score   support

         0.0       0.98      0.99      0.98       995
         1.0       0.98      0.99      0.99      1121
         2.0       0.96      0.97      0.96      1015
         3.0       0.98      0.95      0.97      1033
         4.0       0.97      0.97      0.97       976
         5.0       0.97      0.97      0.97       884
         6.0       0.98      0.98      0.98       999
         7.0       0.98      0.98      0.98      1034
         8.0       0.96      0.95      0.96       923
         9.0       0.95      0.96      0.96      1020

    accuracy                           0.97     10000
   macro avg       0.97      0.97      0.97     10000
weighted avg       0.97      0.97      0.97     10000


[BernoulliRBM] Iteration 1, pseudo-likelihood = -140.21, time = 11.17s
[BernoulliRBM] Iteration 2, pseudo-likelihood = -117.96, time = 15.70s
[BernoulliRBM] Iteration 3, pseudo-likelihood = -105.89, time = 15.29s
[Bernoul

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
C:\Users\dem0nakos\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   18.0s finished


Model performance:
              precision    recall  f1-score   support

         0.0       0.98      0.98      0.98       995
         1.0       0.99      0.99      0.99      1121
         2.0       0.97      0.98      0.98      1015
         3.0       0.97      0.96      0.97      1033
         4.0       0.98      0.97      0.97       976
         5.0       0.97      0.97      0.97       884
         6.0       0.99      0.98      0.98       999
         7.0       0.98      0.98      0.98      1034
         8.0       0.96      0.96      0.96       923
         9.0       0.95      0.96      0.95      1020

    accuracy                           0.97     10000
   macro avg       0.97      0.97      0.97     10000
weighted avg       0.97      0.97      0.97     10000




Στο παραπάνω πρόγραμμα χρησιμοποιούνται τρία διαφορετικά μοντέλα για την ταξινόμηση του συνόλου δεδομένων *MNIST*, όπου είναι μια μεγάλη βάση δεδομένων με χειρόγραφα ψηφία που χρησιμοποιείται συνήθως για την εκπαίδευση διαφόρων συστημάτων επεξεργασίας εικόνας. Τα τρία μοντέλα είναι 1) Η λογιστική παλινδρόμηση χωρίς τη χρήση *RBN*, 2) η λογιστική παλινδρόμηση με τη χρήση *RBM* και η 3) η λογιστική παλινδρόμηση με *DBN* από 2 *RBM*. 

Όπως βλέπουμε και στις τρεις περιπτώσεις η ακριβεία βρίσκεται σε πλήρη αντιστοιχία ως προς τις μετρικές *accuracy, macro average* και *weighted average*. Στον παρακάτω πίνακα συνοψίζονται τα αποτελέσματα.

| Μοντέλο | Ακρίβεια |
| :---: | :---: |
| Λογιστική παλινδρόμηση χωρίς RBM | **92%** |
| Λογιστική παλινδρόμηση με RBM | **97%** |
| Λογιστική παλινδρόμηση με DBN από 2 RBM | **97%** |

Όπως βλέπουμε σε όλες τις περιπτώσεις η ακρίβεια ξεπερνάει το $90\%$. Στη περίπτωση της λογιστική παλινδρόμησης έχουμε ακρίβεια $92%$ ενώ όπως βλέπουμε στις δύο περιπτώσεις που αξιοποιούνται τα *RBM* και τα *DBN* η ακρίβεια ανεβαίνει στο $97\%$.